src/current/advisories/a131639.md

```diff
@@ -0,0 +1,101 @@
+
+While transactions will typically be routed from gateway nodes to the true leaseholder (on the node with the healthy store), and write intents will be written on this true leaseholder, on the node with the slow store these write intents are still queued up in the range's Raft log, and have not been applied. If async intent resolution (the means by which intents are committed to durable records, once a transaction commits) is subsequently triggered on this slow node before it has caught up on its Raft log, intent resolution becomes a no-op, instead of marking the intent as committed. A future transaction that interacts with the row may then come across this intent, and if the committed transaction's record has been garbage collected at that point, it may incorrectly determine that the intent belongs to an aborted transaction. This would then cause the intent to be cleaned up, effectively dropping the committed write.
+
+This bug is rare in that it requires intermittent slowness of a block device backing a store, coupled with lease transfers onto the store with the intermittently slow block device, followed by async intent resolution on the slow node before it has caught up on its Raft log. Writes must touch more than one range. Single range writes, including those that benefit from the One-Phase-Commit (1PC) fast path are not affected.
+
+All CockroachDB versions prior to and including the following are impacted by this bug:
+
+- v23.1.0 - v23.1.26 and below
+- v23.2.0 - v23.2.10 and below
+- v24.1.0 - subsequent patch versions are unaffected
+
+Versions prior to 23.1 are no longer eligible for [maintenance support]({% link releases/release-support-policy.md %}), and this issue will not be addressed in those versions.
+
+## Statement
+
+[#123442](https://github.com/cockroachdb/cockroach/pull/123442) resolves an issue in CockroachDB in the expiration-to-epoch lease promotion transition, where the lease's effective expiration could be allowed to regress, resulting in two nodes believing they are the leaseholder for a range.
+
+The patch has been applied to maintenance releases of CockroachDB:
+
+- [v23.1.27]({% link releases/v23.1.md%}#v23-1-27)
+- [v23.2.11]({% link releases/v23.2.md%}#v23-2-11)
+- [v24.1.1]({% link releases/v24.1.md%}#v24-1-1)
+
+This public issue is tracked by [131639](https://github.com/cockroachdb/cockroach/issues/131639).
+
+## Mitigation
+
+Users are encouraged to upgrade to [v23.1.27]({% link releases/v23.1.md%}#v23-1-27), [v23.2.11]({% link releases/v23.2.md%}#v23-2-11), [v24.1.1]({% link releases/v24.1.md%}#v24-1-1), or a later version that includes the patch.
+
+### Detection via logs
```

Comment:
We now have a [script](https://github.com/cockroachlabs/support/pull/3101) that can be used, given a `debug.zip` as input.

The script requires two dependencies - `ripgrep` and `duckdb`. The install instructions for each can be found here:
- https://github.com/BurntSushi/ripgrep#installation
- https://duckdb.org/docs/installation

@rmloveland - thoughts on how to best include this? We could include the script inline, perhaps at the bottom of the TA, so as not to clutter things to much? Alternatively, we could host it somewhere, and reference it. Will defer to you.

Reply:
@nicktrav (and @nvanbenschoten as script author) I took a shot at including the script inline, and updated the section with an edited version of the description of https://github.com/cockroachlabs/support/pull/3101

PTAL

one thing i noticed is that the script outputs range IDs and the next section 'Detection via an audit of primary and secondary indexes' refers to checking indexes by name

is there guidance we can offer to users on how to get from the potentially affected range IDs to the names of indexes they need to check? 

AFAICT it's pretty easy to get from a schema object to a range ID with [`SHOW RANGES`](https://www.cockroachlabs.com/docs/v24.2/show-ranges.html) but i'm not sure if we have a supported (or unsupported) way to map from range IDs up to schema objects?  if we don't write it here, it seems folks will reach out to ask

Reply:
found some discussion [here](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1716323894389819) of how to do approx a range ID -> SQL mapping and it looks pretty "internal" / unsupported / confusing but i defer to yall

Reply:
This is a good question. Shaun M closes that thread with "The fact that a range can contain primary indexes, secondary indexes, and even have multiple tables within the same range - there’s a many-to-one relationship which is rough to untangle". That's pretty much right, and so it's not as clear an answer as it might seem to perform this mapping, at least not in an automated way. Sean C includes the kind of query we'd need to run to perform the mapping, but it's involved.

If we get to this point with a customer where the script detecting symptoms is pointing at range IDs, it might just be best to get support involved. The next steps are going to be different on a case-by-case basis.

Reply:
Thanks, @nvanbenschoten. Given the various callouts we have to contact Support, can we leave the copy as-is, or is there anything you'd add to point them in the right direction about what to expect, even if the end result is needing to contact Support?

@nickelnickel can you also review this, please?

Reply:
@nvanbenschoten: Nick (@nickelnickel) tells me he's good with this as-is, so if the copy's good to you, we are good to go.

Reply:
I'm good with this as is.